### PR TITLE
Pin Roslyn version for generators in .NET 7 ref pack

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Build out infrastructure for pinned Roslyn dependencies for generators that ship in .NET 7 and update our non-pinned Roslyn dependencies to at least the pinned version. Add some comments so the reasoning behind each pinned version is explained and when we can update them is explained.

This PR does not pin the version used by generators in the repo that do not ship outside the repo, as those generators would not cause build breaks in upstream repos.

cc: @dotnet/interop-contrib this PR pins the version of Roslyn that we compile LibraryImportGenerator against so we can ship in the ref pack without breaking other repos. This is a high enough Roslyn version to have all of the APIs that we're using, so we should be okay.

### CODE REVIEW

**Review by ViktorHofer — APPROVED** [review]

File: src/libraries/System.Text.RegularExpressions/gen/System.Text.RegularExpressions.Generator.csproj
Comment (joperezr) [inline src/libraries/System.Text.RegularExpressions/gen/System.Text.RegularExpressions.Generator.csproj:21]: Won't this mean that the regex source generator wouldn't work any longer in VS today? Correct me if I'm wrong but I believe VS doesn't yet have 4.2.x Roslyn on it, and because this generator is on by default and will run for any application, this would potentially just break creating new projects in VS. cc @ericstj

Comment (ericstj) [reply src/libraries/System.Text.RegularExpressions/gen/System.Text.RegularExpressions.Generator.csproj:21]: @jaredpar @marcpopMSFT will know for sure.  I double checked VS2022 and after upgrading to 17.2 preview 2.1 I see it has a 4.2.0 build of Roslyn so it may be OK with latest.

Comment (jkoritzinsky) [reply src/libraries/System.Text.RegularExpressions/gen/System.Text.RegularExpressions.Generator.csproj:21]: According to the release notes for 7.0.100 preview 2, the latest preview for VS2022 has a new enough Roslyn

Comment (joperezr) [reply src/libraries/System.Text.RegularExpressions/gen/System.Text.RegularExpressions.Generator.csproj:21]: I wonder if we should be conservative just in case, especially when we don't yet need any new features from Roslyn. Main advantage that I see is that if folks are using an older VS2022 preview but with a new SDK, they would still be able to run.

Comment (jkoritzinsky) [reply src/libraries/System.Text.RegularExpressions/gen/System.Text.RegularExpressions.Generator.csproj:21]: We need the 4.2.0 APIs for LibraryImportGenerator, so we need to pin that to this version at minimum when we put in in the ref pack anyway. Also, if we support running on an older VS preview with a newer SDK, then we should say that's the case in the download/release notes. If feel that requiring people who use preview builds to stay up to date isn't a bad thing, especially if our final release requirements will be even higher than these.

Comment (jkoritzinsky) [reply src/libraries/System.Text.RegularExpressions/gen/System.Text.RegularExpressions.Generator.csproj:21]: If we really feel strongly about it, we can hold back the Regex generator, but I don't think it's worthwhile since we're planning on adding the LibraryImport generator soon too.

**Comment by jkoritzinsky** [comment]
@joperezr Any more feedback on this PR? I've got some other work that is quite simplified once this change is in, so I'd like to merge this soon.

File: src/libraries/System.Text.RegularExpressions/gen/System.Text.RegularExpressions.Generator.csproj
Comment (jkoritzinsky) [reply src/libraries/System.Text.RegularExpressions/gen/System.Text.RegularExpressions.Generator.csproj:21]:
```suggestion
    <PackageReference Include="Microsoft.CodeAnalysis.CSharp.Workspaces" Version="$(MicrosoftCodeAnalysisVersion_4_0)" PrivateAssets="all" />
```

File: eng/Versions.props
Comment (sharwell) [inline eng/Versions.props:33]: ❔ Did we ever resolve the issue where these are getting loaded in 16.11?

File: eng/Versions.props
Comment (sharwell) [inline eng/Versions.props:46]:
```suggestion
    <MicrosoftCodeAnalysisCSharpWorkspacesVersion>$(MicrosoftCodeAnalysisVersion_4_X)</MicrosoftCodeAnalysisCSharpWorkspacesVersion>
    <MicrosoftCodeAnalysisCSharpVersion>$(MicrosoftCodeAnalysisVersion_4_X)</MicrosoftCodeAnalysisCSharpVersion>
```

File: eng/Versions.props
Comment (sharwell) [inline eng/Versions.props:48]:
```suggestion
    <MicrosoftCodeAnalysisVersion>$(MicrosoftCodeAnalysisVersion_4_X)</MicrosoftCodeAnalysisVersion>
```

File: eng/Versions.props
Comment (sharwell) [inline eng/Versions.props:44]: 📝 This should be tied to the same value we use for Microsoft.Net.Compilers.Toolset

File: src/libraries/System.Text.RegularExpressions/gen/System.Text.RegularExpressions.Generator.csproj
Comment (joperezr) [reply src/libraries/System.Text.RegularExpressions/gen/System.Text.RegularExpressions.Generator.csproj]: > We can hold off on RegexGenerator for now, but we will have this exact issue when LibraryImportGenerator is included in the ref pack (as it will be on by default as well), which is likely going to happen in a matter of weeks.

Right, I think the main issue here is, the moment we make this change (for the RegexGenerator for now only) then folks using older previews with new SDKs will be automatically broken, and at least as of today, RegexGenerator doesn't necessarily need this change yet. Making this change on the LibraryImportGenerator won't cause this yet, since as you pointed out, it is not yet part of the ref pack. Basically my thinking here, is that we should delay this break until we need it, and when we do, we should notify @marcpopMSFT so that they can make sure that the version we will depend on in the ref pack will be the one present in VS. At that point, we will have no choice but to break folks on older previews but we would need to do so because we depended on the new compiler so I think it would be better justified.

**Review by joperezr — APPROVED** [review]
LGTM from the RegexGenerator side, thanks for undoing the changes on it.

File: eng/Versions.props
Comment (jkoritzinsky) [reply eng/Versions.props:46]: These property values will float separately from MicrosoftCodeAnalysisVersion_4_X very soon (I'm picking up a Roslyn fix for another PR), so I don't want to tie these together just to break the connection. The non-versioned properties are meant for internal-repo usage, so we don't need them to be locked down to particular versions of Roslyn.

**Comment by jkoritzinsky** [comment]
Failures are https://github.com/dotnet/runtime/issues/30724 and https://github.com/dotnet/runtime/issues/67675

**Pull request merged**
